appsrc/dcfile/dcinfo.man
.TH DCINFO 1 "22 August 2026" "DICOM PS3" "DICOM PS3 - Summarize information about DICOM file content"
.SH NAME
dcinfo \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Summarize information about DICOM file content
.SH SYNOPSIS
.HP 10
.B dcinfo
.so man1/gen.so
[
.B \-v|verbose
]
[
.B \-filename
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B dcinfo
reads the named dicom or acr-nema input file and provides summary information about the content of the composite instance,
such as whether or not it is an image, is multi-frame, is enhanced family, is encapsulated and the presence or absence of
offset tables.
.LP
.SH OPTIONS
The description and verbose output go to standard error.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading and once read.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dcinfo XAImage.dcm
.RE
Transfer Syntax to read data set is encapsulated
.RE
Is a non-enhanced family instance
.RE
Is not an instance of a Concatenation
.RE
Is an image with multiple frames
.RE
Is an image with encapsulated PixelData
.RE
Basic Offset Table is present
.RE
Extended Offset Table is absent
.RE

.RE
% dcinfo WSImage.dcm
.RE
Transfer Syntax to read data set is encapsulated
.RE
Is an enhanced family instance
.RE
Is an enhanced family instance without PerFrameFunctionalGroupsSequence
.RE
Has DimensionOrganizationType of TILED_SPARSE
.RE
Has DimensionIndexSequence
.RE
Is not an instance of a Concatenation
.RE
Is an image with multiple frames
.RE
Is an image with encapsulated PixelData
.RE
Basic Offset Table is absent
.RE
Extended Offset Table is present
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR andump(1) ,
.BR dcdump(1) ,
.BR dcfile(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (c) 1993-2026, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
260803:	libsrc/standard/elmdict/other.tpl: Add Agfa private data elements encountered

260729:	libsrc/standard/elmdict/other.tpl: Update and add Sectra private data elements from PACS and VNA Workstation and Core CS 28.1, March 2026

260728:	libsrc/standard/elmdict/other.tpl: Add Infinitt private data elements encountered

260716:	libsrc/standard/elmdict/other.tpl: Add more PixelMed private data elements for TIFF conversion (bugs.mrmf 001474)

260710:	libsrc/src/generic/datetype.cc: Fix memory leak and mmm creation logic (APY e-mail 2026/07/10)

260702:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check for TILED_FULL that number of frames matches total expected if Concatenation (000642)

260702:	libsrc/standard/strval/base.tpl: Add 99EVIDENT CSD encountered

260701:	libsrc/standard/strval/base.tpl: Add 99PRAMANA CSD encountered

260627:	libsrc/standard/elmdict/other.tpl: Add more Pramana WSI private data elements encountered

260624:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check lengths of Extended Offset Table matches NumberOfFrames (000641)

260624:	appsrc/dcfile/dcinfo.cc: Correct detection of empty Extended Offset Table (000640)

260624:	appsrc/dcfile/dcinfo.cc: Report whether instance of concatenation or not (000639)

260616:	libsrc/standard/condn.tpl: Do not complain about presence of Referenced Series Sequence in Common Instance Reference Module when images referenced from Conversion Source Attributes Sequence (000638)

260613:	libsrc/standard/condn.tpl: Check whether or not IOD permits Segmented Lookup Table Data for other than images (000637)

260613:	libsrc/standard/condn.tpl: Do not complain about presence of Referenced Series Sequence in Common Instance Reference Module when images referenced from Advanced Blending Sequence (000636)

260603:	libsrc/standard/elmdict/dicom3.tpl: Add Display URI data element (CP 2595)

260324:	libsrc/standard/elmdict/siemens.tpl: Remove duplicate MR private data elements and add some from XA60A conformance statement and encountered images

260324:	libsrc/standard/condn.tpl,module/vl.tpl: Add Acquisition and Pyramid UID requirements to WG26SP2025 profile (CP-LAB-279)

260324:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Report invalid presence of PyramidUID for LABEL and OVERVIEW image types (000577)

260321:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for metal artifact reduction (CP 2575)

260228:	libsrc/standard/module/base.tpl,vl.tpl: Detect missing PrimaryAnatomicStructureSequence in WG26SP2025 profile (000635)

260228:	Imakefile,dcfile/Imakefile,appsrc/dcfile/dccsdump.cc,dccsdump.man,dccidump.man,dcdump.man: Add utility to dump code sequence items (000634)

260228:	appsrc/dcfile/dccidump.cc,dcsrdump.cc: Show Long Code Value or URN Code Value used when Code Value not present (000633)

260228:	appsrc/dcfile/dccidump.cc: Show numeric values in non-SR content items (NUMERIC rather than NUM) (000632)

260201:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for new defined terms for Dose Type to RT Dose Module (CP 2535)

260123:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Add mechanism for handling sensitive images (CP 1982)

260117:	libsrc/standard/condn.tpl,sopcl.tpl,iodcomp,module,strval/base.tpl: Add DICOM Encapsulation of OBJ Models for 3D Manufacturing and Virtual Reality (Sup 208)

251221:	libsrc/standard/condn.tpl: Check for SCT human species as well as SRT (000630)

251020: appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Warn when DimensionIndexPointer Attribute not present in any Shared or Per-Frame Functional Group (000629)

251020: libsrc/standard/condn.tpl,module/base.tpl: Check DimensionOrganizationType is absent as alternative to not TILED_FULL for DimensionIndexSequence condition (CP 1822)

250914:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Add Scan Spot Gantry and Patient Support Angle to Ion Beam Record (CP 2504)

250907:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Add Dose Reporting Mode and Dose Calculation Model to RT Dose (CP 2428)

250817: libsrc/standard/condn.tpl: Do not report missing PlanePositionSequence and PlaneOrientationSequence for WSI Parametric Map (000628)

250817: libsrc/standard/strval/base.tpl,ct.tpl,mr.tpl: Update Pixel Contrast Image/Frame Type Value 4 to use QUANTITY, expecting details to be in RWVM Quantity Definition Sequence (CP 1700)

250723:	appsrc/dcfile/dcinfo.cc,.man,Imakefile: Add utility to describe characteristics of images esp. re. multi-frame, enhanced, encapsulated and presence or absence of offset tables (000627)

250715:	appsrc/dcfile/dciodvfy.cc: Check consistency of ImagedVolumeWidth,Height with TotalPixelMatrixColumns,Rows and PixelSpacing (000626)

250713:	libsrc/standard/elmdict/other.tpl: Add PixelMed original file message digest private data elements

250709:	libsrc/standard/module/rt.tpl: Relax Type 1 RT Structure Set requirements (CP 2445)

250709:	libsrc/standard/elmdict/base.tpl: Add data element for Add RT specific plan approvals (CP 2427)

250605:	libsrc/standard/transyn.tpl,strval/base.tpl: Add HTJ2K Transfer Syntaxes and string values (Sup 235)

250605:	libsrc/standard/transyn.tpl,strval/base.tpl: Add JPEG-XL Transfer Syntaxes and string values (Sup 232)

250605:	libsrc/include/dctool/vr.h,src/dctool/attrmxrd.cc: Check VR is OB for encapsulated compressed PixelData (000624)

250525:	appsrc/dcfile/dciodvfy.cc,dciodvfy.man,libsrc/src/dctool/attrmxls.cc,libsrc/support/module.awk: Check only first item of Referenced Image Navigation Sequence by default (000623)

250524:	appsrc/dcfile/dciodvfy.cc,dciodvfy.man,libsrc/include/dctool/attrmxls.h,iodcomp.h,module.h,libsrc/src/dctool/attrmxls.cc,libsrc/support/iodcomp.awk,module.awk: Check only first item of Per-Frame Functional Groups Sequence by default, without specific command line option to do them all (000621)

250523:	libsrc/standard/strval/base.tpl: Add 99Visiopharm coding scheme


appsrc/dcfile/dcinfo.cc
static const char *CopyrightIdentifier(void) { return "@(#)dcinfo.cc Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#include "attrmxls.h"
#include "elmconst.h"
#include "mesgtext.h"
#include "dcopt.h"

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool showfilename=options.get("filename");

	dicom_input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);

	cerr << dicom_input_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-v|-verbose]"
			<< " [-filename]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);

	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		log << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	if (verbose) log << "******** While reading ... ********" << endl;
//clock_t start = clock();
	list.read(din,false/*newformat*/,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);
//log << "Elapsed time " << (double((clock() - start))/CLOCKS_PER_SEC*1000.0) << " ms" <<endl;
	
	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}

	if (verbose) log << "******** As read ... ********" << endl;

	// derived from dciodvfy.checkOffsetTables()

	//Attribute *aTransferSyntaxUID=list[TagFromName(TransferSyntaxUID)];
	//char *vTransferSyntaxUID = NULL;
	//if (aTransferSyntaxUID) {
	//	(void)aTransferSyntaxUID->getValue(0,vTransferSyntaxUID);
	//}
	TransferSyntax *dts=din.getTransferSyntaxToReadDataSet();
	if (dts) {
		log << "Transfer Syntax to read data set is " << (dts->getEncapsulated() ? "" : "un") << "encapsulated" << endl;
	}

	Attribute *aSharedFunctionalGroupsSequence=list[TagFromName(SharedFunctionalGroupsSequence)];
	Attribute *aPerFrameFunctionalGroupsSequence=list[TagFromName(PerFrameFunctionalGroupsSequence)];
	if (aSharedFunctionalGroupsSequence
	 || aPerFrameFunctionalGroupsSequence
	 ) {
		log << "Is an enhanced family instance" << endl;
	}
	else {
		log << "Is a non-enhanced family instance" << endl;
	}

	if (aPerFrameFunctionalGroupsSequence) {
		log << "Is an enhanced family instance with PerFrameFunctionalGroupsSequence" << endl;
	}
	else if (aSharedFunctionalGroupsSequence) {
		log << "Is an enhanced family instance without PerFrameFunctionalGroupsSequence" << endl;
	}

	// (000650)
	{
		Attribute *aDimensionOrganizationType=list[TagFromName(DimensionOrganizationType)];
		char *vDimensionOrganizationType = NULL;
		if (aDimensionOrganizationType) {
			(void)aDimensionOrganizationType->getValue(0,vDimensionOrganizationType);
			log << "Has DimensionOrganizationType of " << vDimensionOrganizationType << endl;
		}

		Attribute *aDimensionIndexSequence=list[TagFromName(DimensionIndexSequence)];
		if (aDimensionIndexSequence) {
			log << "Has DimensionIndexSequence" << endl;
		}
	}
	
	// (000639)
	// Do not check whether is also enhanced family, as it should be, but could be standard extended SOP Class
	Attribute *aConcatenationFrameOffsetNumber=list[TagFromName(ConcatenationFrameOffsetNumber)];
	Attribute *aConcatenationUID=list[TagFromName(ConcatenationUID)];
	Attribute *aSOPInstanceUIDOfConcatenationSource=list[TagFromName(SOPInstanceUIDOfConcatenationSource)];
	Attribute *aInConcatenationNumber=list[TagFromName(InConcatenationNumber)];
	Attribute *aInConcatenationTotalNumber=list[TagFromName(InConcatenationTotalNumber)];
	if (aConcatenationFrameOffsetNumber
	 || aConcatenationUID
	 || aSOPInstanceUIDOfConcatenationSource
	 || aInConcatenationNumber
	 || aInConcatenationTotalNumber
	 ) {
		log << "Is an instance of a Concatenation" << endl;
	}
	else {
		log << "Is not an instance of a Concatenation" << endl;
	}

	Attribute *aPixelData=list[TagFromName(PixelData)];

	if (!aPixelData) {
		log << "Is not an image (no PixelData in top level dataset)" << endl;
	}
	else {
		Attribute *aNumberOfFrames=list[TagFromName(NumberOfFrames)];
		Uint32 vNumberOfFrames = 1;
		if (aNumberOfFrames) {
			(void)aNumberOfFrames->getValue(0,vNumberOfFrames);
		}
		if (vNumberOfFrames > 1) {
			log << "Is an image with multiple frames" << endl;
		}
		else {
			log << "Is an image with a single frame" << endl;
		}
		
		if (!aPixelData->isEncapsulated()) {
			log << "Is an image with unencapsulated PixelData" << endl;
		}
		else {
			log << "Is an image with encapsulated PixelData" << endl;
		
			Uint32 offsetOfFirstFragmentItemTag = 0;	// need this for checking both the Basic Offset Table and the Extended Offset Table
			
			BinaryInputStream in(*(istream *)input_opener,LittleEndian);
			
			long offset = aPixelData->getByteOffset();
			//cerr << "checkOffsetTables(): aPixelData offsetToStartOfPixelDataAttribute = " << offset << endl;
			// can assume Explicit VR LE representation since all encapsulated TS are that
			offset += 12;	// length of PixelData group, element, VR and padding, VL
			offset += 4;	// length of Item Tag group, element
				
			in.seekg(offset);	// need to seek from beginning, since may not have been rewound from previous reads
			Uint32 lengthOfBasicOffsetTableInBytes = in.read32();
			//cerr << "checkOffsetTables(): lengthOfBasicOffsetTableInBytes = " << lengthOfBasicOffsetTableInBytes << endl;
			offset+=4;
			
			// check Basic Offset Table, if present ... (000531)
			if (lengthOfBasicOffsetTableInBytes) {
				log << "Basic Offset Table is present" << endl;
			}
			else {
				log << "Basic Offset Table is absent" << endl;
			}

			// check Extended Offset Table, if present ...
			//cerr << "checkOffsetTables(): looking for ExtendedOffsetTable attribute" << endl;
			Uint32 nExtendedOffsetTable = 0;
			const Uint64 *vExtendedOffsetTable = NULL;
			Attribute *aExtendedOffsetTable = list[TagFromName(ExtendedOffsetTable)];
			if (aExtendedOffsetTable) {
				Uint32 lengthOfExtendedOffsetTableInBytes = aExtendedOffsetTable->getVL();	// (000640
				if (lengthOfExtendedOffsetTableInBytes > 0) {
					log << "Extended Offset Table is present" << endl;
				}
				else {
					log << "Extended Offset Table is present but empty" << endl;
				}
			}
			else {
					log << "Extended Offset Table is absent" << endl;
			}
		}
	}
}
